pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -152,6 +159,8 @@ func (t *TopologyGroup) Hash() uint64 {
 	}, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true}))
 }
 
+// nextDomainTopologySpread returns a scheduling.Requirement declaring the node domain that a pod should be scheduled into.
+// If we are unable to find a node domain then we return a "DoesNotExist" node selector
```

Comment:
@jonathan-innis can you help me work out a descriptive and (importantly!) behaviorally correct func comment for this?

In particular, in the `minDomain == ""` outcome where we return a requirement formulated from "DoesNotExist" node selector input, what scheduling requirement are we actually declaring?

Reply:
My understanding is that when `minDomain == ""`, we return an empty requirement which should cause us to throw an error for an `UnsatisfiableTopologySpreadConstraint` on [L161](https://github.com/aws/karpenter-core/blob/main/pkg/controllers/provisioning/scheduling/topology.go#L161), so it's effectively an error for not being able to satisfy a `toplogySpreadConstraint`, given the known domains

Reply:
I'd say

```go
nextDomainTopologySpread returns a scheduling.Requirement selecting a node domain that a pod should be scheduled to. If there are multiple eligible domains, given the `maxSkew`, we return a random domain. If there are no eligible domains, we return a `DoesNotExist` requirement, implying that we could not satisfy the topologySpread requirement
```

Reply:
I think I can definitely see some spots where we can evaluate if the newly supported `minDomains` configuration can't be fulfilled (and thus similarly to `DoesNotExist` express a requirement that is essentially an error), but it's not clear where in this flow a `minDomains` input should have the outcome of scheduling a new node (or nodes) in a new domain (or domains) in order to make forward progress towards a replicaset's eventual consistency of having the minimum required pod distribution across domains.

Reply:
My understanding is that the `t.domainMinCount` should return 0 so long as the number of podDomains with at least a single node in the domain is less than the `minDomains` value. This should have the desired effect where we only consider domains that don't have any pods already scheduled (since the skew would equal 2 if it did) and we'll continue selecting empty domains until we fulfill the `minDomains` requirement

Reply:
> My understanding is that the t.domainMinCount should return 0 so long as the number of podDomains with at least a single node in the domain is less than the minDomains

Reply:
So in this sense the scope of this change is limited to being able to recognize when karpenter *can't* fulfill declared `minDomains` topologySpreadConstraint?

Because I think there is an opportunity to have karpenter create a net new domain as a way of automatically scaling (across the domains dimension) and fulfilling such criteria on behalf of a replicaset (just like it fulfills eventual consistency for replicasets along the "replicas" dimension).

If that's more ambitious work then I agree having a PR that simply fills the gap of `minDomains` being an entirely ignored topologySpreadConstraint property makes sense.

Thoughts?

Reply:
> So in this sense the scope of this change is limited to being able to recognize when karpenter can't fulfill declared minDomains topologySpreadConstraint?

I think so, I can't think of a normal situation where minDomains affects Karpenter unless you're trying to do something weird like a topology spread across CAS & Karpenter.

> Because I think there is an opportunity to have karpenter create a net new domain as a way of automatically scaling (across the domains dimension) and fulfilling such criteria on behalf of a replicaset (just like it fulfills eventual consistency for replicasets along the "replicas" dimension).

We do that already if you have a requirement with an Exists operator, see https://karpenter.sh/v0.27.3/concepts/scheduling/#exists-operator . minDomains won't affect hostname spreads for Karpenter as we assume we can always launch another node.  For AZ spreads, it doesn't make sense to set minDomains to more than the number of AZs, and we know those anyway.  

The situation that minDomains really helps with isn't with Karpenter, but with the kube-scheduler. You can let the scheduler know that you expect something to spread across 3 AZs, even if you only have a single node in your cluster. If you use the DoNotSchedule setting, it will mark the pod as unschedulable and then Karpenter has a chance to launch capacity.  Without this feature, the scheduler will pack as many of the pods onto the single node that already exists before marking any as unschedulable.

We should implement it to be correct and potentially enable some advanced use cases, but I think for the most part minDomains won't affect Karpenter too much.  Its more about the scheduler -> autoscaler interaction.

Reply:
@tzneal We spread across custom domains too, right? I think I recall if the user were to define a set of custom label keys with a set of custom-values (say in the case of [virtual domains](https://karpenter.sh/v0.27.3/concepts/scheduling/#on-demandspot-ratio-split)) that we will automatically spread those evenly?

Reply:
> We spread across custom domains too, right? I think I recall if the user were to define a set of custom label keys with a set of custom-values (say in the case of [virtual domains](https://karpenter.sh/v0.27.3/concepts/scheduling/#on-demandspot-ratio-split)) that we will automatically spread those evenly?

Yes, but we know what those domains are and how many, so minDomains shouldn't affect that. Unless you are spreading across something else like another scale-to-zero autoscaler like CAS.  Maybe there's another case where it would matter, but I couldn't think of one.

> Also, how does the Exists operator work with a topologySpreadConstraint. I don't think I've ever tried it, so curious what the interactions would be with minDomains in that case where we are constructing arbitrary domains.

Good question, we should treat that like hostnames where we can create a new domain on the fly by launching a node. I don't know if we do though.